src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs

```diff
@@ -51,24 +50,26 @@ public long Estimate(Transaction tx, BlockHeader header, EstimateGasTracer gasTr
                 ? tx.GasLimit
                 : header.GasLimit;
 
-            UInt256 senderBalance = _stateProvider.GetBalance(tx.SenderAddress);
-
-            // Calculate and return additional gas required in case of insufficient funds.
-            if (tx.Value != UInt256.Zero && tx.Value >= senderBalance)
+            // Execute binary search to find the optimal gas estimation.
+            try
             {
-                return gasTracer.CalculateAdditionalGasRequired(tx, releaseSpec);
+                return BinarySearchEstimate(leftBound, rightBound, tx, header, cancellationToken)
+                       + Math.Max(0, gasTracer.CalculateAdditionalGasRequired(tx, releaseSpec));
+            }
+            catch (OperationCanceledException)
+            {
+                return long.MaxValue;
```

Comment:
Shouldn't we let the exception bubble-up so the JSON RPC can handle it and return time-out to the user? If we return 0 it will get confusing?

Reply:
I agree, I think letting the exception bubble up is the correct decision, but I was unsure about how we handle these kind of timeouts.

I'll make the change to let the exception bubble up.